cloud/scope/powervs_cluster.go

```diff
@@ -1166,6 +1206,390 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
 	return subnetDetails.ID, nil
 }
 
+// ReconcileVPCSecurityGroup reconciles VPC security group.
+func (s *PowerVSClusterScope) ReconcileVPCSecurityGroup() error {
+	for _, securityGroup := range s.IBMPowerVSCluster.Spec.VPCSecurityGroups {
+		var securityGroupID *string
+		var securityGroupRuleIDs []*string
+
+		if securityGroup.Name != nil {
+			securityGroupID, securityGroupRuleIDs = s.GetVPCSecurityGroupID(*securityGroup.Name)
+		} else {
+			securityGroupRuleIDs = s.GetVPCSecurityGroupRuleIDs(*securityGroup.ID)
+		}
+
+		if securityGroupID != nil && securityGroupRuleIDs != nil {
+			if _, _, err := s.IBMVPCClient.GetSecurityGroup(&vpcv1.GetSecurityGroupOptions{
+				ID: securityGroupID,
+			}); err != nil {
+				return err
+			}
+			for _, rule := range securityGroupRuleIDs {
+				if _, _, err := s.IBMVPCClient.GetSecurityGroupRule(&vpcv1.GetSecurityGroupRuleOptions{
+					SecurityGroupID: securityGroupID,
+					ID:              rule,
+				}); err != nil {
+					return err
+				}
+			}
+			continue
+		}
+
+		securityGroupDet, ruleIDs, err := s.validateVPCSecurityGroup(securityGroup)
+		if err != nil {
+			return fmt.Errorf("error validating existing security group: %v", err)
+		}
+		if securityGroupDet != nil {
+			s.Logger.Info("Security group already exists", "name", *securityGroupDet.Name)
+			s.SetVPCSecurityGroupID(*securityGroupDet.Name, infrav1beta2.VPCSecurityGroupStatus{
+				ID:                securityGroupDet.ID,
+				RuleIDs:           ruleIDs,
+				ControllerCreated: ptr.To(false),
+			})
+
+			continue
+		}
+
+		securityGroupID, err = s.createVPCSecurityGroup(securityGroup)
+		if err != nil {
+			return fmt.Errorf("error creating security group: %v", err)
+		}
+		ruleIDs, err = s.createVPCSecurityGroupRules(securityGroup, securityGroupID)
+		if err != nil {
+			return fmt.Errorf("error creating security group rules: %v", err)
+		}
+		s.SetVPCSecurityGroupID(*securityGroup.Name, infrav1beta2.VPCSecurityGroupStatus{
+			ID:                securityGroupID,
+			RuleIDs:           ruleIDs,
+			ControllerCreated: ptr.To(true),
+		})
+	}
+
```

Comment:
Can we log a statement on successful creation of a security group?

Reply:
Done